app/models/host/discovered.rb

```diff
@@ -84,10 +86,23 @@ def attributes_to_import_from_facts
     super
   end
 
-  def populate_fields_from_facts(facts = self.facts_hash, ignored_type = nil)
-    # detect interfaces and primary interface using extensions
-    parser = super(facts, :foreman_discovery)
+  def populate_fields_from_facts(*params)
+    if SETTINGS[:version].version > '1.16'
+      parser, type, source_proxy = params
+      facts = parser.facts
+      # detect interfaces and primary interface using extensions
+      super(parser, type, source_proxy)
+    else
+      # backwards compatibility
+      facts = params[0] || self.facts_hash
```

Comment:
If there are only 3 parameters I'd rather see another parameter with a default in the signature here and see what each parameter is then use params[0].

Reply:
Previous signature was 
``` ruby
populate_fields_from_facts(facts = self.facts_hash, type = 'puppet')
```
It was 2 parameters with completely different meaning. For example in 1.17 is mandatory and it's the parser, but in 1.16 it was facts hash and was optional. So now I mimic an optional parameter - use passed value and if it does not exist, use default (`self.facts_hash`) the same should go for the second parameter, but it's ignored anyway.

Reply:
Thanks for clarifying.